src/mame/heathkit/h89.cpp

```diff
@@ -291,16 +336,57 @@ void h89_state::h89_io(address_map &map)
 	map(0xe8, 0xef).rw(m_console, FUNC(ins8250_device::ins8250_r), FUNC(ins8250_device::ins8250_w));
 
 	// ports defined on the H8. On the H89, access to these addresses causes a NMI
-	map(0xf0, 0xf1).rw(FUNC(h89_state::raise_NMI_r),FUNC(h89_state::raise_NMI_w));
+	map(0xf0, 0xf1).rw(FUNC(h89_base_state::raise_NMI_r),FUNC(h89_base_state::raise_NMI_w));
 
 	// General Purpose Port (GPP)
-	map(0xf2, 0xf2).w(FUNC(h89_state::port_f2_w)).portr("SW501");
+	map(0xf2, 0xf2).w(FUNC(h89_base_state::port_f2_w)).portr("SW501");
 
 	// port defined on the H8. On the H89, access to these addresses causes a NMI
-	map(0xfa, 0xfb).rw(FUNC(h89_state::raise_NMI_r), FUNC(h89_state::raise_NMI_w));
+	map(0xfa, 0xfb).rw(FUNC(h89_base_state::raise_NMI_r), FUNC(h89_base_state::raise_NMI_w));
+}
+
+void h88_state::h88_io(address_map &map)
+{
+	h89_base_state::h89_base_io(map);
+
+	// Cassette I/O (uses 0xf8 - 0xf9) - Requires MTR-88 ROM
+	map(0xf8, 0xf9).rw(m_cassette, FUNC(heath_h_88_cass_device::read), FUNC(heath_h_88_cass_device::write));
 }
 
+void h89_state::h89_io(address_map &map)
+{
+	h89_base_state::h89_base_io(map);
+
+	// H37 5-1/4" Soft-sectored Controller - Requires MTR-90 ROM
+	map(0x78, 0x7b).rw(m_h37, FUNC(heath_z37_fdc_device::read), FUNC(heath_z37_fdc_device::write));
+}
+
+
 // Input ports
+static INPUT_PORTS_START( h88 )
+
+	PORT_START("SW501")
+	// MTR-88  (444-40)
+	PORT_DIPNAME( 0x1f, 0x00, DEF_STR( Unused ) )               PORT_DIPLOCATION("SW501:1,2,3,4,5")
+	PORT_DIPSETTING( 0x00, "Undefined" )
```

Comment:
The convention is to have individual switches for the unused DIP switches.  You can use `PORT_DIPUNUSED_DIPLOC` to reduce the number of lines a bit, e.g.
```c++
	PORT_DIPUNUSED_DIPLOC(0x01, 0x00, "SW501:1")
	PORT_DIPUNUSED_DIPLOC(0x02, 0x00, "SW501:2")
	PORT_DIPUNUSED_DIPLOC(0x04, 0x00, "SW501:3")
	PORT_DIPUNUSED_DIPLOC(0x08, 0x00, "SW501:4")
	PORT_DIPUNUSED_DIPLOC(0x10, 0x00, "SW501:5")
```